toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -0,0 +1,179 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/builtin_function_kind.h"
+
+#include <utility>
+
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::SemIR {
+
+// A function that validates that a builtin was declared properly.
+using ValidateFn = auto(const File& sem_ir, llvm::ArrayRef<TypeId> arg_types,
+                        TypeId return_type) -> bool;
+
+namespace {
+// Information about a builtin function.
+struct BuiltinInfo {
+  llvm::StringLiteral name;
+  ValidateFn* validate;
+};
+
+// The maximum number of type parameters any builtin needs.
+constexpr int MaxTypeParams = 1;
+
+// State used when validating a builtin signature that persists between
+// individual checks.
+struct ValidateState {
+  // The type values of type parameters in the builtin signature. Invalid if
+  // either no value has been deduced yet or the parameter is not used.
+  TypeId type_params[MaxTypeParams] = {TypeId::Invalid};
+};
+
+// Constraint that a type is generic type parameter `I` of the builtin,
+// satisfying `TypeConstraint`. See ValidateSignature for details.
+template <int I, typename TypeConstraint>
+struct TypeParam {
+  static_assert(I >= 0 && I < MaxTypeParams);
+
+  static auto Check(const File& sem_ir, ValidateState& state, TypeId type_id)
+      -> bool {
+    if (state.type_params[I].is_valid() && type_id != state.type_params[I]) {
+      return false;
+    }
+    state.type_params[I] = type_id;
+    return TypeConstraint::Check(sem_ir, state, type_id);
+  }
+};
+
+// Constraint that requires the type to be an integer type. See
+// ValidateSignature for details.
+struct AnyInt {
+  static auto Check(const File& sem_ir, ValidateState& /*state*/,
+                    TypeId type_id) -> bool {
+    if (sem_ir.types().GetInstId(type_id) == InstId::BuiltinIntType) {
+      return true;
+    }
+    // TODO: Support iN for all N, and the Core.BigInt type we use to implement
+    // for integer literals.
+    return false;
+  }
+};
+}  // namespace
+
+// Validates that this builtin has a signature matching the specified signature.
+//
+// `SignatureFnType` is a C++ function type that describes the signature that is
+// expected for this builtin. For example, `auto (AnyInt, AnyInt) -> AnyInt`
+// specifies that the builtin takes values of two integer types and returns a
+// value of a third integer type. Types used within the signature should provide
+// a `Check` function that validates that the Carbon type is expected:
+//
+//   auto Check(const File&, ValidateState&, TypeId) -> bool;
+//
+// To constrain that the same type is used in multiple places in the signature,
+// `TypeParam<I, T>` can be used. For example:
+//
+// `auto (TypeParam<0, AnyInt>, AnyInt) -> TypeParam<0, AnyInt>`
```

Comment:
Consistency suggestion:

```suggestion
// a `Check` function that validates that the Carbon type is expected:
//
// `auto Check(const File&, ValidateState&, TypeId) -> bool;`
//
// To constrain that the same type is used in multiple places in the signature,
// `TypeParam<I, T>` can be used. For example:
//
// `auto (TypeParam<0, AnyInt>, AnyInt) -> TypeParam<0, AnyInt>`
```

or

```suggestion
// a `Check` function that validates that the Carbon type is expected:
//
//   auto Check(const File&, ValidateState&, TypeId) -> bool;
//
// To constrain that the same type is used in multiple places in the signature,
// `TypeParam<I, T>` can be used. For example:
//
//   auto (TypeParam<0, AnyInt>, AnyInt) -> TypeParam<0, AnyInt>
```